src/lib/logging.ts

```diff
@@ -0,0 +1,453 @@
+        const level = this.levelToInt(logLevel);
+        if (level < this.minLevel) return;
+
+        const logEntry: LogEntry = {
+            level: logLevel,
+            module: this.module,
+            message,
+            objects,
+            traceId: undefined,
+        };
+
+        // Obtain the line/file through a thoroughly hacky method
+        // This creates a new stack trace and pulls the caller from it.  If the caller
+        // if .trace()
+        // const error = new Error('');
+        // if (error.stack) {
+        //     const cla = error.stack.split('\n');
+        //     let idx = 1;
+        //     while (idx < cla.length && cla[idx].includes('at Logger.Object.')) idx++;
+        //     if (idx < cla.length) {
+        //         logEntry.location = cla[idx].slice(cla[idx].indexOf('at ') + 3, cla[idx].length);
+        //     }
+        // }
+
+        this.logManager.emit('log', logEntry);
+    }
+
+    public trace(message: string, objects?: any): void {
+        this.log('trace', message, objects);
+    }
+    public debug(message: string, objects?: any): void {
+        this.log('debug', message, objects);
+    }
+    public info(message: string, objects?: any): void {
+        this.log('info', message, objects);
+    }
+    public warn(message: string, objects?: any): void {
+        this.log('warn', message, objects);
+    }
+    public error(message: string, objects?: any): void {
+        this.log('error', message, objects);
+    }
+
+    /**
+     * Central logging method with a trace ID to track calls between modules/components.
+     * @param logLevel
+     * @param message
+     */
+    public logWithId(
+        logLevel: string,
+        traceId: string,
+        message: string,
+        objects?: any,
+    ): void {
+        const level = this.levelToInt(logLevel);
+        if (level < this.minLevel) return;
+
+        const logEntry: LogEntry = {
+            level: logLevel,
+            module: this.module,
+            message,
+            objects,
+            traceId,
+        };
+
+        this.logManager.emit('log', logEntry);
+    }
+
+    public traceWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('trace', traceId, message, objects);
+    }
+    public debugWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('debug', traceId, message, objects);
+    }
+    public infoWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('info', traceId, message, objects);
+    }
+    public warnWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('warn', traceId, message, objects);
+    }
+    public errorWithId(traceId: string, message: string, objects?: any): void {
+        this.logWithId('error', traceId, message, objects);
+    }
+}
+
+type TimingMap = {
+    // count, avg, min, max
+    [id: string]: number[];
```

Comment:
What does that comment mean? Is there a reason why those 4 names are not represented in code?

Reply:
I was calculating metrics on call times to debug some performance issues. This is a simple JS hashmap where the id could be one of the 4 items.

Reply:
OK I don't really understand that, but I'll a comment to the code so the info is not lost.